### DISCUSSION TITLE
Add more Vector Properties

### DISCUSSION THREAD

**Discussion by donn-xx** [godotengine/godot-proposals#7986] [Scripting]
Vector types have x,y,z properties but that limits the readability of code when you want to hold things like:

- height, width, depth [h,w,d]
- left, up, forwards [l,u,f]

Code jazz
```
var size := Vector3(1,2,5) # w,h,d
if size.w > 1: dostuff()
size.h *= 10
etc.
```

I have found myself using Dictionaries or Classes instead because the properties are explicit and help me think. With Vectors it's all x,y,z even when those letters do not apply to the problem.

It would be even cooler if we could *add* properties to the Vectors, like:

```
var v := Vector2
v.add_prop("volume", "v")
v.add_prop("amplitude", "a")
```
Thought?

**Comment by Zireael07** [comment]
That would make them no longer vectors but structs. See https://github.com/godotengine/godot-proposals/issues/7329 for a structs proposal

**Reply by donn-xx** [reply]
Thanks. I see they are going back and forth about Vectors becoming structs. I'll leave it up to the boffins.

**Reply by Zireael07** [reply]
Haha, yep, last I saw that discussion was a week ago and I see that it seems to be going in circles now :P

**Comment by theraot** [comment]
> left, up, forwards [l,u,f]

I think you mean right, up, backwards :P

---

Sometimes I find myself wanting to alias the vectors.

And I do mean alias them. Not new properties, not any change to the underlying type, nor custom vectors. Just syntactic sugar that let me call the vector type and its components something else in a particular script.

The reason is that in lack of <strike>tuples</strike> (errata: typed tuples or typed dictionaries, or what Juan Linietsky calls "structs", or similar), I have used vectors in unusual ways.

I have no idea how much trouble this would cause for the parser. And I'm not convinced alias as I imagine them should take any priority over existing proposals.

**Reply by donn-xx** [reply]
That's a neat idea. Fully agree.

**Reply by donn-xx** [reply]
> I think you mean right, up, backwards
Lol. You just demonstrated a real problem with my suggestion :D

**Reply by aXu-AP** [reply]
In context of gdscript, what would be the advantage of tuples over dictionaries? Syntax-wise dictionaries are already quite easy on eyes.

	var box = { width = 5, height = 3, depth = 8 }
	print(box.depth)

**Reply by theraot** [reply]
Amended.
